### PR TITLE
CryptoArena application: zencode-for-substrate.md

### PR DESCRIPTION
**Pull request by Irvine**
# Grant Application Checklist

- [x] [application-template.md](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been duplicated, renamed `zencode-for-substrate.md`, and was updated.
- [x] An Ethereum (DAI) address for the payment of the milestones is provided inside the application.  
- [ ] The software of the project will be released under the Apache license version 2.0 as specified in the [terms and conditions](https://gist.github.com/Noc2/75bc58e8ce9b5d419ff883b0cf2b8c19). (We received verbal permission from [Parity Technologies](https://www.parity.io/) to select GNU GPL v3 due to its compatibility with dependencies.)
- [x] The total funding amount of the project is below USD $30k for initial grant applications and $100k for follow-up grants.
- [x] The initial PR contains only one commit.
- [x] The grant will only be announced once we've successfully [delivered](https://github.com/w3f/Grant-Milestone-Delivery) the first milestone.

### CODE REVIEW

**Comment by semuelle** [comment]
Hi @Irvine, thanks for your application. We will have an editor check your document for completeness next week, who will share it with the rest of the team after.

File: applications/zencode-for-substrate.md
Comment (semuelle) [inline applications/zencode-for-substrate.md:8]: Do you mean, this application is in response to a Request For Proposals? If so, could you add a link? Or what do you mean by RFP?

File: applications/zencode-for-substrate.md
Comment (semuelle) [inline applications/zencode-for-substrate.md:103]: Just so we are on the same page: this application is for Milestone 1 only, correct? It's good to know that you have concrete plans for the future, though. :)

File: applications/zencode-for-substrate.md
Comment (CryptoArenaCEO) [reply applications/zencode-for-substrate.md:8]: Hi @semuelle, I read somewhere in the instructions we had to specify if it was a request for proposal - by which we intend a document to solicit a request, from you to us, to build the proposed milestone. 

Could be a misunderstanding on my part if you mean something else.

File: applications/zencode-for-substrate.md
Comment (CryptoArenaCEO) [reply applications/zencode-for-substrate.md:103]: Correct, just Milestone 1 for now, and we'll apply for follow ups in due time. 
And thank you, it'll be cool!

File: applications/zencode-for-substrate.md
Comment (semuelle) [reply applications/zencode-for-substrate.md:8]: I see. Well, since you already know what you want built and you want to implement it yourself, this would be considered a grant application. RFPs can indeed by written by anyone, but they are intended to attract developers, who can then submit grant applications in reference to an RFP. So, "proposals" are grant applications, and RFPs are "invitations for grant applications" for a specific topic. RFPs lack the most specific information, such as the name of a/the team implementing it (see our [RFP template](https://github.com/w3f/General-Grants-Program/blob/master/rfp-proposal/suggestion-template.md)).

Comment (CryptoArenaCEO) [reply applications/zencode-for-substrate.md:8]: Understood. Thanks for the disambiguation, we'll amend it to say its a grant application

Comment (Irvine) [reply applications/zencode-for-substrate.md]: @semuelle and @CryptoArenaCEO, I've updated the application to indicate that it is a grant application instead of an RFP.

File: applications/zencode-for-substrate.md
Comment (CryptoArenaCEO) [reply applications/zencode-for-substrate.md:103]: Hi @semuelle , can I ask you how long should we expect the review process to take? 
We're eager & ready.

Comment (semuelle) [reply applications/zencode-for-substrate.md:103]: Hi @CryptoArenaCEO, if you don't hear from one of us this week, please complain. 
Generally, an application needs three approvals, meaning three evaluators need to read, understand and approve the application. We have no strict timelines or deadlines on these; depending on the number of questions and complexity of the application, it might take a few days to several weeks.

**Comment by Irvine** [comment]
@semuelle, I believe this should address all of the points in your previous comment. All responses were written by @CryptoArenaCEO unless mentioned. Please let us know if you'd like any additional details.

1. How: Zencode allows plain English smart contracts. Its interpreter Zenroom is secure, efficient, scalable and written in C, thus will run reliably on basically anything. This means its potential application is vast and crosses business sectors, specializations and geography. You can make a smart contract to control your car's airbags if you can find a use for it. Here are some different examples:
- CryptoArena aims to include it as part of our copy trading modules, to enable zk non-custodial peer to peer trade contracts
- A milestone beyond number 2 that we came up but didn't list would be a decentralized supply chain management / procurement solution
- Legal solutions (more advanced applications)
- A decentralized school system to check homework submission/correctness and award merit points

Why: It removes the barrier of a programming language-like syntax, from the usage of smart contracts. Anyone can use it, or even customize it for own purposes, even when lacking any and all technical knowledge. It could reach many who wouldn’t touch a smart contract otherwise, and even familiarize them with the tech.

2. Zenroom is an independent interpreter of the smart contract that provides a secure environment for fundamental cryptographic operations that Zencode performs. Milestone 1 is what we will integrate into the pallet, namely a dedicated API "engine" to call Zenroom functions from, process input and translate it to the specified recipe – substrate. The communication flow will be: CLI -> API (Zenroom wrapping) -> Zenbridge -> Substrate Pallet

3. The command line interface is just a comfortable interface to call functions from; those functions are the fundamental cryptographic operations that zencode performs. [Zencode Advanced](https://dev.zenroom.org/#/pages/zencode-crypto) examples include public key cryptography / signature, zk attribute-based credentials, or petitions.

4. In summary, the zk-SNARKs are used by Coconut's authentication flow... which is also used for some advanced proofs in Zenroom called [scenarios](https://dev.zenroom.org/#/pages/zencode-scenario-credentials). This allows two parties that wish to perform a transaction to perform it safely, by means of two generated zk proxies which verify each other's hashes. This comes into play in our future solution by enabling secure and anonymous non-custodial copy trading.

5. @Irvine: Adding a new commit to fix this now; apologies for the typos.

6. Prioritization of this work has been pending confirmation of the grant. Related links:
- [Documentation](https://dev.zenroom.org/#/pages/zencode-cookbook-intro)
- [Downloads & derivatives](http://zenroom.org/#downloads)

Of the above, we plan to use [Zenbridge](https://zenroom.org/zenbridge/) for the first milestone. Other useful repositories:
- [Zenroom](https://github.com/dyne/Zenroom)
- [Social Wallet API](https://github.com/dyne/social-wallet-api): our future plans, out of scope for the proposed grant work

@Irvine and @rabm2k7: As we already chose Node.js to create modules to integrate Zenroom with Substrate, the UI will be a command line interface that leverages Node.js modules such as [clui](https://github.com/nathanpeck/clui) to construct a menu-driven interface for deployment, [minimist](https://www.npmjs.com/package/minimist) to parse arguments, and [inquirer](https://www.npmjs.com/package/inquirer) to provide interactivity with sane default values. Providing arguments such as `--identityfile` and `--chain` upon command execution will allow for bypassing menu options to automate deployment, balancing a friendly UI with a powerful one. We'll keep it simple to match the natural language nature of Zenroom smart contracts, and grow it as we add features.

**Comment by semuelle** [comment]
Hi all, thanks for the follow-up. Generally, I am quite interested in the idea of natural language smart contracts, but at the same time the setup you have envisioned seems a bit convoluted. Anyway, I'll let the committee decide. Just two more questions:

- What is a "smart contract recipe"?
- Can you explain in a bit more detail your own use case, i.e. the copy trading? Would users write and deploy their own Zencode smart contracts locally, or do you do it for them?

**Comment by CryptoArenaCEO** [comment]
> Hi all, thanks for the follow-up. Generally, I am quite interested in the idea of natural language smart contracts, but at the same time the setup you have envisioned seems a bit convoluted. Anyway, I'll let the committee decide. Just two more questions:

Hi All, hi @semuelle , thanks for your interest, replies & time. 
A bit convoluted work for us, yes, but it's a tool meant to make things very easy for others once its done, both end-Users and other Devs extending it for their own purposes.

> * What is a "smart contract recipe"?

Recipes are like plugins for the interpreter (Zenroom) of the English-like language (Zencode).
Basically, they are blockchain specific compatibility modules. 

The recipe based smart contract engine takes in the instructions, performs the cryptographic operations that are unique to the specified chain according to the "recipe" and broadcasts the resulting transaction / performs the required function.  

These allow the Zenchain Pallet blockchain ( = Milestone 2 ) to sign blocks in such a way that is compatible with the target blockchain. 
Using the Polka Relay Chain as example: sign valid blocks that'll go through collators & validators friction-less-ly.
 
New recipes can be created by developers to make Zenchain target other specific blockchains (existing or future), thus allowing that blockchain to use English like smart contracts too. Further, existing “recipes” can be loaded & deployed with ease, through the CLI interface.

> * Can you explain in a bit more detail your own use case, i.e. the copy trading? Would users write and deploy their own Zencode smart contracts locally, or do you do it for them?

They write/deploy it locally. 
From their perspective its just 3 simple steps with no code involved. Select, negotiate, finalize. 

Pictures help:
User_Perspective

More info: 
- the two special user categories are referred to as "Patrons" (Copier) & "Champion" (Copied)
- the unique terms of their interaction are referred to as the "patronage", and the smart contract that is deployed as a consequence of their negotiated agreement is referred to as a "Patronage agreement"
- a patronage agreement ("PA") is essentially a quid pro quo trading contract between pseudonymous strangers.
- each PA requires at least 3 conditions to be present and agreed to by both parties in order to be valid
These are: a) a Target RoI, b) the Champion's Fee, c) Termination/withdrawal conditions.
- On top of these basics, Users have many customization options. All of these settings are set by them, in english.
- The only condition we impose from our end on all PAs is: Success Fees basis. 
Champion don't get paid their fee unless they return a profit to their Patron, specifically a profit larger than the sum of the champion's fee + 1/2 the same fee.

Even more info? check out this link to our (work in progress) docs:
https://cryptoarena-1.gitbook.io/docs/-MWdPS9e2j4WJYQcFOIS/

Then navigate to "service offering" submenu. 

Always available for more info as may be necessary.

**Comment by semuelle** [comment]
Hi all, thanks for the quick follow-up. I appreciate the effort.

I'm going to share your application with the rest of the committee. In the meantime, could you integrate the info you gave above into the document?

**Comment by CryptoArenaCEO** [comment]
> Hi all, thanks for the quick follow-up. I appreciate the effort.
> 
> I'm going to share your application with the rest of the committee. In the meantime, could you integrate the info you gave above into the document?

No problem :)  
@Irvine can you take care of updating the doc please?

Looking forward to the committee's reply.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. Looks like an interesting project. However, it sounds like the “Zencode Smart Contract Engine” is off-chain (second layer?). How does it ensure the correct execution of a “smart contract”? Milestone 1 seems to be almost unrelated to polkadot at the moment. Doesn’t it make more sense to start with milestone 2 instead?

**Comment by CryptoArenaCEO** [comment]
> Thanks for the application. Looks like an interesting project. 

Hi @Noc2, thanks.  

> However, it sounds like the “Zencode Smart Contract Engine” is off-chain (second layer?). 

@rabm2k7 do chime in to correct me if I say something wrong anywhere.

Yes, engine is off chain. It's kinda like what Remix is for ethereum, a web based IDE. Flow being servers -> trusted node -> blockchain. We probably should call it Zenchain IDE or something along those lines to avoid confusion.

> How does it ensure the correct execution of a “smart contract”? 

The IDE doesn't **ensure** it itself, albeit it'll come with a test environment to try things out. It is the consensus protocol of the chain that ensures the production of valid blocks to feed to the relay chain. 

Which protocol? Eventually all of the ones included in Substrate. Initially? We'd like to talk to about it further with you guys. 
For our own purposes it'll (likely) be a custom flavor of PoA.

It is possible to deploy broken contracts, which will result in invalid blocks that won't be accepted, which implies the loss of the cost of deployment with no working code in return (a net gain for the network overall). 
This is one of the rationales that contributes to the value of the project - as milestone two will be working template pallet blockchain pre-built for compatibility with Polkadot, to use zencode with that produces valid blocks. They can just clone & build upon further (as we intend to ourselves for the purpose of copy trading).

> Milestone 1 seems to be almost unrelated to polkadot at the moment. Doesn’t it make more sense to start with milestone 2 instead?

Well, they're both development & deployment tooling, one is an IDE, the other a parachain development kit. Milestone 1 we will use to deploy milestone 2, and also it includes the recipe for compatibility with Polka.

I do agree milestone 2 is the "big target", but I would say they're very much related ( and we'd be more than happy to commit to both, as I had mentioned in our emails some time back ). To some degree we're adjusting to fit the format and expectations of the grant program.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the answer. But to be honest, I would only be interested in an application if it initially focuses on the on-chain integration of your smart contract language. This could for example be via the compilation to WebAssembly, see [Contract Module](https://github.com/paritytech/substrate/tree/master/frame/contracts). Milestone 1 in its current stage could be completely useless, if milestone 2 is unsuccessful and milestone 2 is currently very unspecific. As you pointed out yourself, you currently don’t know how and if milestone 2 is going to be implemented.

**Comment by alxs** [comment]
I agree with David's last comment and my suggestion would be to simply apply at a later date. I'm also sceptical on natural language programming in itself, as I think some programming knowledge will always be needed in order to put together a smart contract i.e. in the example provided in the Zenroom introductory article

> Given that I am known as 'Citizen'
and I have my keypair
and I have a signed credential
and I use the verification key by 'State Digital Authority'
When I aggregate all the verification keys
and I sign the petition 'To not build the wall'
Then print all data

if a user knows what this all means, then they likely have some basic programming skills and can just as well develop a smart contract, and if they don't they won't succeed using Zencode either. It seems dangerous to me to claim that someone with no coding skills can deploy a smart contract, no matter the level of abstraction.

Nevertheless, I think NL contracts could find some use mostly in trading bots, which is the use case you're already covering and all in all it still looks like an interesting project to me, so I'd be happy to consider an alternative application as long as it focuses on Substrate/Polkadot integration.

**Comment by CryptoArenaCEO** [comment]
Hi Guys, @Noc2 @alxs thanks.

It really shouldn't be too much of a problem to start with M2, it's just a bit less comfortable schedule, but we can do it.
We can deploy a basic transactional blockchain pallet to demonstrate the functionality of zencode, and add more fancy later.

> Thanks for the answer. But to be honest, I would only be interested in an application if it initially focuses on the on-chain integration of your smart contract language. This could for example be via the compilation to WebAssembly, see Contract Module. Milestone 1 in its current stage could be completely useless, if milestone 2 is unsuccessful and milestone 2 is currently very unspecific. As you pointed out yourself, you currently don’t know how and if milestone 2 is going to be implemented.

It's not the bad kind of uncertainty. We have options and can do it in several ways according to what's best for existing systems, if there are preferences from its side, which we're unsure of.

> I agree with David's last comment and my suggestion would be to simply apply at a later date. I'm also sceptical on natural language programming in itself, as I think some programming knowledge will always be needed in order to put together a smart contract i.e. in the example provided in the Zenroom introductory article
> 
> > Given that I am known as 'Citizen'
> > and I have my keypair
> > and I have a signed credential
> > and I use the verification key by 'State Digital Authority'
> > When I aggregate all the verification keys
> > and I sign the petition 'To not build the wall'
> > Then print all data
> 
> if a user knows what this all means, then they likely have some basic programming skills and can just as well develop a smart contract, and if they don't they won't succeed using Zencode either. It seems dangerous to me to claim that someone with no coding skills can deploy a smart contract, no matter the level of abstraction.

It's a bit of a fantastic claim its true, but also simpler than it sounds. 
let me start with a disambiguation: this example above that you pasted is a **Petition scenario**. 

image

This is actually among advanced proofs using ZK-snark authentication flow that I had mentioned in a previous comment.
That's indeed not for common users.

image

They can however access the Basics with ease. Link to the above indicated docs section is this: https://dev.zenroom.org/#/pages/zencode-cookbook-given

These work with the combination of **three** simple types of clauses that anyone can understand: **"given", "when", "then".**

example:

> > Given that the w3f committee approves our application,
> > When we deliver a cool tool for NL smart contracts,
> > Then the committee will evaluate if it does what it should,
> > When the committee approves milestone delivery
> > Then the disbursement of our grant payment shall be processed.

Anyone can do these - and it gets much simpler/better by building a specific and limited framework of textual clause options that the users can combine through a GUI. 

Basically, it's simple to package it and to provide it "rails" for a specific purpose (as we intend to do for copy trading) so that all possible combinations of the options available to users return either a working smart contract or an error that prevents them from deploying it.


image

> Nevertheless, I think NL contracts could find some use mostly in trading bots, which is the use case you're already covering and all in all it still looks like an interesting project to me, so I'd be happy to consider an alternative application as long as it focuses on Substrate/Polkadot integration

We have: 
- expansions to this project,
- a gas tracker app coming soon to android market, only for ETH costs at the moment
- our main platform for which we aim to get into the substrate chains builder program (and to lease a parathread/parachain). Find a ton of info about this here: https://cryptoarena-1.gitbook.io/docs/-MWdPS9e2j4WJYQcFOIS/

**Comment by CryptoArenaCEO** [comment]
> I just realised Zencode/Zenroom is not your own work and your dev team consists of one rather junior developer (no offense) who's also your CTO + one frontend. Besides, the two GitHub profiles you provide under past experience are empty and CryptoArena is closed-source, as far as I can tell. To be honest I don't think I would support this even if we were talking about a pallet.
> 
> Nevertheless, feel free to update this application as suggested above and the committee will decide.

This is borderline slander, and we've done nothing to deserve it. 

- Its Dyne's under gnu v3 and we've credited them repeatedly, and have been nothing but transparent. 
In fact we originally reached out proposing cryptoarena itself, and were advised to start with this project by others in your organizations, specifically so that we could demonstrate our quality to you.
 
- Cryptoarena is not closed source, its just early, and as a regulated platform has to go through auditing first. It will be all open source by the time we get to market, and we even have a 100% non profit mission.
- We are indeed talking about a Pallet, 
- I have 3 -we added one- full stack engineers, one and half of which with blockchain, and a front end,
- My devs are plenty capable, in spite of your superficial judgement based on social profiles. 
- And most importantly the committee is in control of delivery checks, if the work we submit is not up to your standards, you refuse it and lose nothing. 

The new proposal is: everything scrapped except pallet (formerly "milestone 2") for the 30k grant, if you want it say so and we'll amend the application. Otherwise, I'm done jumping through hoops for you.

**Comment by semuelle** [comment]
Just to clarify, @CryptoArenaCEO: I don't think @alxs was insinuating that you are infringing licenses, but realised that Zenroom is a different team. Which is a relevant point in the evaluation of the application.

Regarding your suggestion, I think it is in line with @Noc2's opinion. I certainly would like to see it happen.

**Comment by CryptoArenaCEO** [comment]
@alxs @semuelle I appreciate your replies, and I'm not really the type for grudges.

Further, I noticed the reader link I shared to our gitbook still required you to be "added" to our organization or sign in, which may be source of misunderstanding. So, apologies from me for this. I've set them to public temporarily for you to read freely: https://docs.cryptoarena.org

There's plenty of interesting info there about our plans for cryptoarena, such as an explanation as to how our distribution chain would eventually combine probabilistic finality and deterministic proof to achieve proven finality, for the purpose of signing blocks to send to the relay chain. (See "proof of revenue distribution": https://docs.cryptoarena.org/self-decentralization/how-it-works/pord)

Also please note this is all still under revision, some sections are missing, and in some sentences I state the likes of "cryptoarena is built as a parachain" while the technically correct statement would be "cryptoarena is planned to be built as a parachain". 
Polka is our top choice for L1, but we obviously don't have a leased slot assigned to us as of now.

There is no code in the gitbook yet. Only explanations, albeit with good levels of detail. Code work examples we'll add to the github repository for you to review in the next application. Of our own as CryptoArena, that we can show now, we have:

- our gas tracker app - which is basically complete, in react/golang, for historicals, track custom ranges & set alerts
- the smart contracts for our utility tokens (see CATs in documentation), solidity / truffle
- a bunch of endpoints, actions and logic for our exchange app 

of our devs' past outside CryptoArena, I need to ask them details of what they can share. That I remember they did, a CRM, a wallet, a swap protocol, databases and more. 
Further, we have established friendly relations with other organization either willing or eager to help us if we encounter particular issues.

Revised application will come shortly,
Best,
Lorenzo

**Comment by alxs** [comment]
@CryptoArenaCEO any updates? We generally close application after two weeks of inactivity.

**Comment by CryptoArenaCEO** [comment]
> @CryptoArenaCEO any updates? We generally close application after two weeks of inactivity.

Hi @alxs , everyone.

Apologies for the delay, it is coming. It just took a little longer than we expected to approve, and had to give priority to catching up on other tasks & deadlines as we have our first token sale in 10 days. We should be good to go in no time.

**Comment by semuelle** [comment]
@CryptoArenaCEO: I don't know in what way your grant application and token sale are related, but:

> We do not award grants for projects that have been the object of a successful token sale.

https://github.com/w3f/Open-Grants-Program#guidelines

**Comment by CryptoArenaCEO** [comment]
> @CryptoArenaCEO: I don't know in what way your grant application and token sale are related, but:
> 
> > We do not award grants for projects that have been the object of a successful token sale.
> 
> https://github.com/w3f/Open-Grants-Program#guidelines

Hi @semuelle , it isn't related to the proposed grant work. 
Different entity, different blockchain, different scope. The token sale (https://docs.cryptoarena.org/cat-intro) is to fund part of our unique exchange platform, liquidity for listings on dexes, and an NFT based trading cards game deployed as an Eth DApp. 

No token relating to the "zenchain pallet" or to its future developments is to be minted, either now or later.

**Comment by semuelle** [comment]
I will close this PR for now. Feel free to reopen it once you have an updated application. However, it might make even more sense to separate it from the conversation here and to create new PR. Either way, let us know if you have any updates, @CryptoArenaCEO.

**Comment by CryptoArenaCEO** [comment]
@semuelle I understand & thank you for your patience, we surely will very soon. We restarted work on it earlier today, we're still doing the bits we took out, just not asking for the w3f to fund it.

As update the utility token sale I mentioned last time is going great and still open - we raised 90 eth and growing:
r1: https://etherscan.io/address/0x531018B98ab6f0E32464f2145d4fc1951328f67d (sold in 20 mins)
r2: https://etherscan.io/address/0x7f3cdbfba742c4f06af163f0f95247d898037aa5 (still running)

And we're starting to make waves, we got influencers & youtubers, an hour and a half slot at the Boston Blockchain Week and more. In fact we could use guests, if there any takers among you. Bet you'd ask very interesting questions.

**Pull request closed without merging**
